File: apps/child/src/native/textmodel.ts
export type TextPrediction = { category: string; severity: string; confidence: number };

export type TextModel = {
  categories: string[];
  severities: string[];
  buckets: number;
  predict(text: string): TextPrediction | null;
  raw(text: string): { category: string; severity: string; confidence: number; sevConfidence: number } | null;
};

export const NULL_THRESHOLD = 0.5;
export const HYBRID_THRESHOLD = 0.85;

const MAX_CP = 300;
const FNV_OFFSET = 0x811c9dc5;
const FNV_PRIME = 0x01000193;
const P_WORD = 1;
const P_BIGRAM = 2;
const P_CHAR = 3;
const SPACE = 32;
const TOKEN_RE = /[\p{L}\p{Nd}\p{Nl}\p{No}\p{Mn}]/u;

function fnvByte(h: number, b: number): number {
  return Math.imul(h ^ b, FNV_PRIME) >>> 0;
}

function fnvCp(h: number, cp: number): number {
  if (cp < 0x80) return fnvByte(h, cp);
  if (cp < 0x800) return fnvByte(fnvByte(h, 0xc0 | (cp >> 6)), 0x80 | (cp & 0x3f));
  if (cp < 0x10000) {
    h = fnvByte(h, 0xe0 | (cp >> 12));
    h = fnvByte(h, 0x80 | ((cp >> 6) & 0x3f));
    return fnvByte(h, 0x80 | (cp & 0x3f));
  }
  h = fnvByte(h, 0xf0 | (cp >> 18));
  h = fnvByte(h, 0x80 | ((cp >> 12) & 0x3f));
  h = fnvByte(h, 0x80 | ((cp >> 6) & 0x3f));
  return fnvByte(h, 0x80 | (cp & 0x3f));
}

function isTokenCp(cp: number): boolean {
  if (cp < 0x80) return (cp >= 0x61 && cp <= 0x7a) || (cp >= 0x30 && cp <= 0x39) || (cp >= 0x41 && cp <= 0x5a);
  return TOKEN_RE.test(String.fromCodePoint(cp));
}

function isEmojiCp(cp: number): boolean {
  return (cp >= 0x1f000 && cp <= 0x1faff) || (cp >= 0x2600 && cp <= 0x27bf);
}

type Norm = { cps: number[]; starts: number[]; ends: number[] };

function normalise(text: string): Norm {
  const cps: number[] = [SPACE];
  const starts: number[] = [];
  const ends: number[] = [];
  let inTok = false;
  let n = 0;
  for (const ch of text.toLowerCase()) {
    if (n >= MAX_CP) break;
    n++;
    const cp = ch.codePointAt(0)!;
    if (isTokenCp(cp)) {
      if (!inTok) {
        starts.push(cps.length);
        inTok = true;
      }
      cps.push(cp);
    } else {
      if (inTok) {
        ends.push(cps.length);
        cps.push(SPACE);
        inTok = false;
      }
      if (isEmojiCp(cp)) {
        starts.push(cps.length);
        cps.push(cp);
        ends.push(cps.length);
        cps.push(SPACE);
      }
    }
  }
  if (inTok) {
    ends.push(cps.length);
    cps.push(SPACE);
  }
  return { cps, starts, ends };
}

function hashRange(prefix: number, cps: number[], from: number, to: number, mask: number): number {
  let h = fnvByte(FNV_OFFSET, prefix);
  for (let i = from; i < to; i++) h = fnvCp(h, cps[i]!);
  return h & mask;
}

export function featurize(text: string, buckets: number): Map<number, number> {
  const mask = buckets - 1;
  const { cps, starts, ends } = normalise(text);
  const counts = new Map<number, number>();
  if (starts.length === 0) return counts;
  const bump = (b: number) => counts.set(b, (counts.get(b) ?? 0) + 1);
  for (let i = 0; i < starts.length; i++) bump(hashRange(P_WORD, cps, starts[i]!, ends[i]!, mask));
  for (let i = 0; i + 1 < starts.length; i++) bump(hashRange(P_BIGRAM, cps, starts[i]!, ends[i + 1]!, mask));
  const L = cps.length;
  const seed = fnvByte(FNV_OFFSET, P_CHAR);
  for (let i = 0; i < L; i++) {
    let h = seed;
    for (let k = 0; k < 5 && i + k < L; k++) {
      h = fnvCp(h, cps[i + k]!);
      if (k >= 2) bump(h & mask);
    }
  }
  let ss = 0;
  for (const v of counts.values()) ss += v * v;
  const inv = 1 / Math.sqrt(ss);
  for (const [b, v] of counts) counts.set(b, v * inv);
  return counts;
}

type Head = { n: number; scale: Float32Array; w: Int8Array; bias: Float32Array };

function readHead(dv: DataView, buf: ArrayBuffer, off: number, buckets: number, n: number): { head: Head; off: number } {
  const scale = new Float32Array(n);
  for (let i = 0; i < n; i++) scale[i] = dv.getFloat32(off + 4 * i, true);
  off += 4 * n;
  const w = new Int8Array(buf, off, buckets * n);
  off += buckets * n;
  const bias = new Float32Array(n);
  for (let i = 0; i < n; i++) bias[i] = dv.getFloat32(off + 4 * i, true);
  off += 4 * n;
  return { head: { n, scale, w, bias }, off };
}

function softmaxInto(z: Float64Array): number {
  let m = -Infinity;
  for (let i = 0; i < z.length; i++) if (z[i]! > m) m = z[i]!;
  let s = 0;
  for (let i = 0; i < z.length; i++) {
    z[i] = Math.exp(z[i]! - m);
    s += z[i]!;
  }
  let best = 0;
  for (let i = 0; i < z.length; i++) {
    z[i] = z[i]! / s;
    if (z[i]! > z[best]!) best = i;
  }
  return best;
}

function runHead(head: Head, feats: Map<number, number>): Float64Array {
  const acc = new Float64Array(head.n);
  for (const [b, v] of feats) {
    const base = b * head.n;
    for (let c = 0; c < head.n; c++) acc[c] = acc[c]! + v * head.w[base + c]!;
  }
  for (let c = 0; c < head.n; c++) acc[c] = acc[c]! * head.scale[c]! + head.bias[c]!;
  return acc;
}

export function loadTextModel(buf: ArrayBuffer): TextModel {
  const dv = new DataView(buf);
  const magic = String.fromCharCode(dv.getUint8(0), dv.getUint8(1), dv.getUint8(2), dv.getUint8(3));
  if (magic !== "LHTM") throw new Error("textmodel: bad magic");
  const version = dv.getUint16(4, true);
  if (version !== 1) throw new Error(`textmodel: unsupported version ${version}`);
  const buckets = dv.getUint32(6, true);
  const nClasses = dv.getUint8(10);
  const nSev = dv.getUint8(11);
  let off = 12;
  const dec = new TextDecoder();
  const names: string[] = [];
  for (let i = 0; i < nClasses + nSev; i++) {
    const len = dv.getUint8(off);
    names.push(dec.decode(new Uint8Array(buf, off + 1, len)));
    off += 1 + len;
  }
  const categories = names.slice(0, nClasses);
  const severities = names.slice(nClasses);
  const cat = readHead(dv, buf, off, buckets, nClasses);
  const sev = readHead(dv, buf, cat.off, buckets, nSev);
  const safeIdx = categories.indexOf("SAFE");

  const raw = (text: string) => {
    if (!text) return null;
    const feats = featurize(text, buckets);
    if (feats.size === 0) return null;
    const pc = runHead(cat.head, feats);
    const ps = runHead(sev.head, feats);
    const ci = softmaxInto(pc);
    const si = softmaxInto(ps);
    return { category: categories[ci]!, severity: severities[si]!, confidence: pc[ci]!, sevConfidence: ps[si]! };
  };

  return {
    categories,
    severities,
    buckets,
    raw,
    predict(text: string) {
      const r = raw(text);
      if (!r) return null;
      if (categories.indexOf(r.category) === safeIdx || r.confidence < NULL_THRESHOLD) return null;
      return { category: r.category, severity: r.severity, confidence: r.confidence };
    },
  };
}

File: apps/child/native-src/TextModel.kt
package fund.fastforward.lighthouse.child

import android.content.Context
import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.math.exp
import kotlin.math.sqrt

data class Prediction(val category: String, val severity: String, val confidence: Float)

object TextModel {
  const val NAME = "lh-textmodel-v1"
  private const val ASSET = "textmodel.bin"
  private const val MAX_CP = 300
  private const val FNV_OFFSET = -0x7EE3623B
  private const val FNV_PRIME = 0x01000193
  private const val P_WORD = 1
  private const val P_BIGRAM = 2
  private const val P_CHAR = 3
  private const val SPACE = 32
  const val NULL_THRESHOLD = 0.5f
  const val HYBRID_THRESHOLD = 0.85f

  private class Head(val n: Int, val scale: FloatArray, val w: ByteArray, val wOff: Int, val bias: FloatArray)
  private class Model(
    val buckets: Int,
    val categories: List<String>,
    val severities: List<String>,
    val cat: Head,
    val sev: Head,
    val safeIdx: Int,
  )

  @Volatile private var model: Model? = null

  private val counts = IntArray(1 shl 17)
  private var touched = IntArray(4096)
  private var nTouched = 0
  private var norm = IntArray(2 * MAX_CP + 4)
  private var starts = IntArray(MAX_CP + 2)
  private var ends = IntArray(MAX_CP + 2)

  fun init(ctx: Context) {
    try {
      val bytes = ctx.assets.open(ASSET).use { it.readBytes() }
      model = parse(bytes)
    } catch (_: Exception) {
      model = null
    }
  }

  fun load(bytes: ByteArray): Boolean {
    return try {
      model = parse(bytes)
      true
    } catch (_: Exception) {
      false
    }
  }

  fun isReady(): Boolean = model != null

  private fun parse(bytes: ByteArray): Model {
    val bb = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN)
    if (bytes.size < 12 || bytes[0] != 'L'.code.toByte() || bytes[1] != 'H'.code.toByte() || bytes[2] != 'T'.code.toByte() || bytes[3] != 'M'.code.toByte()) {
      throw IllegalArgumentException("bad magic")
    }
    val version = bb.getShort(4).toInt() and 0xFFFF
    if (version != 1) throw IllegalArgumentException("unsupported version $version")
    val buckets = bb.getInt(6)
    if (buckets <= 0 || buckets and (buckets - 1) != 0 || buckets > (1 shl 20)) throw IllegalArgumentException("bad buckets")
    val nClasses = bytes[10].toInt() and 0xFF
    val nSev = bytes[11].toInt() and 0xFF
    var off = 12
    val names = ArrayList<String>(nClasses + nSev)
    repeat(nClasses + nSev) {
      val len = bytes[off].toInt() and 0xFF
      names.add(String(bytes, off + 1, len, Charsets.UTF_8))
      off += 1 + len
    }
    val categories = names.subList(0, nClasses).toList()
    val severities = names.subList(nClasses, nClasses + nSev).toList()

    fun readHead(n: Int): Head {
      val scale = FloatArray(n) { bb.getFloat(off + 4 * it) }
      off += 4 * n
      val wOff = off
      off += buckets * n
      val bias = FloatArray(n) { bb.getFloat(off + 4 * it) }
      off += 4 * n
      return Head(n, scale, bytes, wOff, bias)
    }
    val cat = readHead(nClasses)
    val sev = readHead(nSev)
    if (off > bytes.size) throw IllegalArgumentException("truncated")
    if (buckets != counts.size) throw IllegalArgumentException("bucket count mismatch")
    return Model(buckets, categories, severities, cat, sev, categories.indexOf("SAFE"))
  }

  private fun fnvByte(h: Int, b: Int): Int = (h xor b) * FNV_PRIME

  private fun fnvCp(h0: Int, cp: Int): Int {
    var h = h0
    if (cp < 0x80) return fnvByte(h, cp)
    if (cp < 0x800) {
      h = fnvByte(h, 0xC0 or (cp shr 6))
      return fnvByte(h, 0x80 or (cp and 0x3F))
    }
    if (cp < 0x10000) {
      h = fnvByte(h, 0xE0 or (cp shr 12))
      h = fnvByte(h, 0x80 or ((cp shr 6) and 0x3F))
      return fnvByte(h, 0x80 or (cp and 0x3F))
    }
    h = fnvByte(h, 0xF0 or (cp shr 18))
    h = fnvByte(h, 0x80 or ((cp shr 12) and 0x3F))
    h = fnvByte(h, 0x80 or ((cp shr 6) and 0x3F))
    return fnvByte(h, 0x80 or (cp and 0x3F))
  }

  private fun isTokenCp(cp: Int): Boolean {
    if (cp < 0x80) return (cp in 0x61..0x7A) || (cp in 0x30..0x39) || (cp in 0x41..0x5A)
    if (Character.isLetter(cp)) return true
    return when (Character.getType(cp).toByte()) {
      Character.DECIMAL_DIGIT_NUMBER, Character.LETTER_NUMBER, Character.OTHER_NUMBER, Character.NON_SPACING_MARK -> true
      else -> false
    }
  }

  private fun isEmojiCp(cp: Int): Boolean = (cp in 0x1F000..0x1FAFF) || (cp in 0x2600..0x27BF)

  private fun bump(b: Int) {
    if (counts[b] == 0) {
      if (nTouched == touched.size) touched = touched.copyOf(touched.size * 2)
      touched[nTouched++] = b
    }
    counts[b]++
  }

  private fun featurize(text: String, mask: Int): Int {
    val lower = text.lowercase()
    var len = 1
    norm[0] = SPACE
    var nTok = 0
    var inTok = false
    var i = 0
    var seen = 0
    while (i < lower.length && seen < MAX_CP) {
      val cp = lower.codePointAt(i)
      i += Character.charCount(cp)
      seen++
      if (isTokenCp(cp)) {
        if (!inTok) {
          starts[nTok] = len
          inTok = true
        }
        norm[len++] = cp
      } else {
        if (inTok) {
          ends[nTok++] = len
          norm[len++] = SPACE
          inTok = false
        }
        if (isEmojiCp(cp)) {
          starts[nTok] = len
          norm[len++] = cp
          ends[nTok++] = len
          norm[len++] = SPACE
        }
      }
    }
    if (inTok) {
      ends[nTok++] = len
      norm[len++] = SPACE
    }
    if (nTok == 0) return 0
    for (t in 0 until nTok) {
      var h = fnvByte(FNV_OFFSET, P_WORD)
      for (k in starts[t] until ends[t]) h = fnvCp(h, norm[k])
      bump(h and mask)
    }
    for (t in 0 until nTok - 1) {
      var h = fnvByte(FNV_OFFSET, P_BIGRAM)
      for (k in starts[t] until ends[t + 1]) h = fnvCp(h, norm[k])
      bump(h and mask)
    }
    val seed = fnvByte(FNV_OFFSET, P_CHAR)
    for (s in 0 until len) {
      var h = seed
      var k = 0
      while (k < 5 && s + k < len) {
        h = fnvCp(h, norm[s + k])
        if (k >= 2) bump(h and mask)
        k++
      }
    }
    return nTouched
  }

  private fun runHead(head: Head, inv: Float, out: FloatArray) {
    val n = head.n
    for (c in 0 until n) out[c] = 0f
    val w = head.w
    for (j in 0 until nTouched) {
      val b = touched[j]
      val v = counts[b] * inv
      val base = head.wOff + b * n
      for (c in 0 until n) out[c] += v * w[base + c]
    }
    for (c in 0 until n) out[c] = out[c] * head.scale[c] + head.bias[c]
  }

  private fun softmax(z: FloatArray, n: Int): Int {
    var m = Float.NEGATIVE_INFINITY
    for (c in 0 until n) if (z[c] > m) m = z[c]
    var s = 0f
    for (c in 0 until n) {
      z[c] = exp((z[c] - m).toDouble()).toFloat()
      s += z[c]
    }
    var best = 0
    for (c in 0 until n) {
      z[c] /= s
      if (z[c] > z[best]) best = c
    }
    return best
  }

  private val catOut = FloatArray(16)
  private val sevOut = FloatArray(16)

  @Synchronized
  fun predict(text: String): Prediction? {
    val m = model ?: return null
    if (text.isEmpty()) return null
    try {
      nTouched = 0
      featurize(text, m.buckets - 1)
      if (nTouched == 0) return null
      var ss = 0L
      for (j in 0 until nTouched) {
        val v = counts[touched[j]].toLong()
        ss += v * v
      }
      val inv = (1.0 / sqrt(ss.toDouble())).toFloat()
      runHead(m.cat, inv, catOut)
      runHead(m.sev, inv, sevOut)
      val ci = softmax(catOut, m.cat.n)
      val si = softmax(sevOut, m.sev.n)
      val conf = catOut[ci]
      if (ci == m.safeIdx || conf < NULL_THRESHOLD) return null
      return Prediction(m.categories[ci], m.severities[si], conf)
    } catch (_: Exception) {
      return null
    } finally {
      for (j in 0 until nTouched) counts[touched[j]] = 0
      nTouched = 0
    }
  }
}
